### PR TITLE
RegSet implementation

### PR DESCRIPTION
**Pull request by nalmt**

### CODE REVIEW

File: lib/src/data_structures.rs [hunk 1]
```diff
@@ -18,6 +18,766 @@ use std::slice::{Iter, IterMut};
 
 use crate::interface::Function;
 
+//=============================================================================
+// RegSet
+
+const BLOCK_SIZE: usize = 32;
+type Block = u32;
```
Comment (nalmt) [inline lib/src/data_structures.rs]: À supprimer

Comment (nalmt) [inline lib/src/data_structures.rs]: À supprimer.

File: lib/src/data_structures.rs [hunk 2]
```diff
@@ -18,6 +18,863 @@ use std::slice::{Iter, IterMut};
...
+      self.data[i] = 0;
+    }
+  }
+
+  fn union(&mut self, other: &Self) {
+    let vec_size = {
+      if other.data.len() <= self.data.len() {
+        other.data.len()
+      } else {
+        for _ in 0..(other.data.len() - self.data.len()) {
```
Comment (moy) [inline lib/src/data_structures.rs:144]: Maybe more natural as

```suggestion
        for _ in self.data.len()..other.data.len() {
```

File: lib/src/data_structures.rs [hunk 3]
```diff
@@ -18,6 +18,863 @@ use std::slice::{Iter, IterMut};
...
+
+    let (first, second) = {
+      if other.data.len() < self.data.len() {
+        (other, self)
+      } else {
+        (self, other)
+      }
+    };
+    for i in 0..first.data.len() {
+      subset = subset && ((self.data[i] | other.data[i]) == other.data[i]);
```
Comment (moy) [inline lib/src/data_structures.rs:191]: You may optimize this with an early return when subset becomes false.

File: lib/src/data_structures.rs [hunk 4]
```diff
@@ -18,6 +18,863 @@ use std::slice::{Iter, IterMut};
...
+    }
+    for i in first.data.len()..second.data.len() {
+      subset = subset && second.data[i] == 0;
```
Comment (moy) [inline lib/src/data_structures.rs:194]: Sounds incorrect to me if `self` is shorter than `other`. In this case, there's nothing to check, the extra part of `other` can't result in a false value.

**Review by moy — CHANGES_REQUESTED** [review]

File: lib/src/data_structures.rs (hunk 4 above)
Comment (moy) [reply lib/src/data_structures.rs:194]: Tip: avoid leaving messages like this unanswered, and make it easy for the reviewer to follow, e.g. by answering "Fixed in #2", I've changed it to ...

(GitHub also has a "resolve conversation" button to keep track of what needs to be answered, and what's not).

Comment (nalmt) [reply lib/src/data_structures.rs:194]: Indeed the returned value is wrong if self is shorter than other, here I added a test case: 
https://github.com/nalmt/regalloc.rs/pull/2/commits/fafcc754e36f7202fa64fed2fe24fe4b4d6bdafe

Fixed here:
https://github.com/nalmt/regalloc.rs/pull/2/commits/1b380873235d38c24707505283a84123087931cc

Thank you!

**Comment by nalmt** [comment]
closed (new pull request opened: https://github.com/nalmt/regalloc.rs/pull/2)

**Pull request closed without merging**